src/Common/ThreadPool.cpp

```diff
@@ -339,12 +342,12 @@ void ThreadPoolImpl<Thread>::worker(typename std::list<Thread>::iterator thread_
             {
                 thread_it->detach();
```

Comment:
Looks like this detach() leads to data race. The only thing that helped to avoid `condition_variable` destruction was mutex. But now we are not holding it anymore, and the thread was detached, so ThreadPool itself could be destroyed after we release mutex and before we call `cv.notify_*()`.

Reply:
Exactly, moving the _job_finished.notify_all()_ out of this critical section is harmful, as the context switch may happen after the _scheduled_jobs_ is decreased to 0 and right before the call of _job_finished.notify_all()_. And the switch-to thread may be the main thread which checks the _scheduled_jobs_ in ThreadPoolImpl::wait() and destruct the thread pool (along with the condition variables) afterwards when it equals 0.